### ISSUE TITLE
Add a token refresh load testing scenario

### ISSUE THREAD

**Issue by ahus1** [keycloak/keycloak-benchmark#416]
### Description

There is currently no scenario to test refreshing access tokens. 

The new suggested Gatling scenario steps or the new scenario are:

1. Login 100_000 users as fast as possible to have a lot of active sessions
2. Refresh access tokens when 60 seconds are left for the access token 

Assumptions: 
* Under peak load, Keycloak will slow down for the responses to the access tokens. As the clients will calculate the next refresh time based on the response, the refreshes will level over time automatically. 

Expected outcomes/learnings: 
* What Keycloak configuration (how many pods/vCPUs) can support this load?
* How does Keycloak behave with many active sessions if we kill pods during the first and second phase?
* Are there bottlenecks?

### Motivation

This is one of the most common activities (even more frequent than user logins), so there should be one for this.

**Comment by ddurham2** [comment]
@ahus1 , I agree this scenario needs to exist.  Browsers left open all refreshing their tokens over time is a realistic use-case.  And just how many users can be doing this before problem start to happen.

I heard one story that since refresh tokens are stored only in the infinispan cache, which is about 10k entries, then users start to be logged out.  True(?) or not, it makes sense if infinispan isn't overflow to disk or db.

So another thing to check for is if the logged in users start to fail to refresh because of this.

Have you started any work on this yourself yet?

**Comment by ddurham2** [comment]
@ahus1, I started a draft PR attempting to address this.  Never done Scala, but I think I'm getting it.  

So far... (links to PR)
1. KeycloakScenarioBuilder.exchangeCode() now [saves](https://github.com/keycloak/keycloak-benchmark/pull/434/files#diff-6cd77690a1aaa2dc4d2aef1b73a8137e47e08345dab512836362aa6a9966070aR247) the refresh token
2. (new method) [KeycloakScenarioBuilder.refreshToken()](https://github.com/keycloak/keycloak-benchmark/pull/434/files#diff-6cd77690a1aaa2dc4d2aef1b73a8137e47e08345dab512836362aa6a9966070aR265) attempts to refresh the id, access and refresh tokens by supplying the refresh token saved when exchangeCode was used.  It POSTs to the same TOKEN_ENDPOINT as exchangeCode()
3. defined [RefreshToken scenario](https://github.com/keycloak/keycloak-benchmark/pull/434/files#diff-3afc410dad746402ab0d29b26735e9d7e99f85a0de8f7de86fe2d6413037718cR11) class which logs in, and then will repeatedly invoke refreshToken()  (not fully implemented)

But I'm having an issue with the refreshToken() routine.  It makes the request, but Keycloak is giving the following HTTP 401 error in the response:
`{"error":"unauthorized_client","error_description":"Invalid client or Invalid client credentials"}`

Keycloak itself logs this:
`2023-07-19 06:41:01,479 WARN  [org.keycloak.events] (executor-thread-119) type=REFRESH_TOKEN_ERROR, realmId=f37169e7-43c1-492d-be41-f26a8f0842d9, clientId=client-0, userId=null, ipAddress=10.244.0.1, error=invalid_client_credentials, grant_type=refresh_token`

The userId is null, which I thought looked odd.

I was thinking this might be due to the cookies set during auth that are marked Secure not being submitted by Gatling since it's not using TLS.  So, for the moment, I [made it re-create](https://github.com/keycloak/keycloak-benchmark/pull/434/files#diff-6cd77690a1aaa2dc4d2aef1b73a8137e47e08345dab512836362aa6a9966070aR253) those cookies without the secure flag, but it still gives the same error.

As far as I can tell, the HTTP POST Gatling is submitting all the same values that my normal application (of course, using a different instance of keycloak and client) does when it successfully refreshes the tokens.

I checked through the `test-realm`'s settings to see if something is disallowing refreshing tokens.  e.g. Token lifetimes > 0, user-0's roles, "Use refresh tokens" set to On, "Use refresh tokens for client credentials grant" set to On, etc.

Any ideas what could be going on?  TIA

**Comment by ddurham2** [comment]
I found the problem to be that client-0 -> Settings -> Capability config -> Client authentication was set to ON

Turning it OFF allows the refresh token operation to work.

**Comment by kami619** [comment]
@ddurham2 thanks for picking up this task to create a new refresh token scenario. I am not fully aware why the publicClient is set to `false` TBH., but that is what we had for the service enabled account, does this have something to do with the cookie's we are setting for a particular session in your code ? Do we really need to set those cookies for this scenario to work ?

**Comment by kami619** [comment]
> But I'm having an issue with the refreshToken() routine. It makes the request, but Keycloak is giving the following HTTP 401 error in the response:

how are you exactly testing this routine ? the Service enabled client expects the correct client-secret and client name for the scenario to work properly. if you are using test-realm the client secret is by default set to `setup-for-benchmark`

**Comment by ddurham2** [comment]
All, now I'm to the point of wanting to actually figure out how this test should work.  Some questions.

- Presumably the tester should be able to specify the number of users (which is already determined by --concurrent-users)

- The tester needs a way to specify the pause between refreshes.  An argument?  Or do we want to just let that be determined by the expiration of the tokens which can be configured in KC admin.  That is, if you want a rapid refresh, then you make the token time, like, 10 seconds, and presumably the refresh operation follows that schedule - w/ 10% remaining or so.  If you want a realistic test, leave it 5m and it'll refresh every 4m30s (i.e. 10% remaining).  The downside is that setting the token time very short may break other (future?) scenarios.

- The test has to end at some point.  Do we have a fixed token refresh count, specified by an arg.  Or should it read the `--mesurement` duration given to the test? And it simply refreshes on the above interval until that time has elapsed?

- The original description talks about 100k users.  I would agree that we need to be testing in that range (or more).  But currently the setup only creates user-0.  I'm going to assume that if we tried to do all the testing with a single user account (logged in 100k times), that it isn't a real-world scenario since db/cache indexes and such aren't really in play like they would be.  So that will require initialize-keycloak-entities.sh to learn how to create N users, and for the test to be told how many users there are.  Correct?

- Finally, I've already discovered with the minimal test that it is, that it is really not hard _at all_ to crash keycloak.  This is probably something that should be discussed in the keycloak project, but I'd be willing to discuss that with somewhere here first if you like.  I'm using a fairly standard setup in minikube.

**Comment by kami619** [comment]
@ddurham2 I am glad you were able to figure out the issue around invalid client credentials issue. And all of your questions related to the scenario are valid and here is my attempt to answer them to the best of my knowledge.

Not sure if you have seen this documentation already, I will be referring it in few places of my answers., https://www.keycloak.org/keycloak-benchmark/benchmark-guide/latest/

> Presumably the tester should be able to specify the number of users (which is already determined by --concurrent-users)

As mentioned in the [configuration](https://www.keycloak.org/keycloak-benchmark/benchmark-guide/latest/configuration) options that you can specify for all the benchmark scenarios., there are two ways to specify users one is the `--concurrent-users`(Closed Model) which you already pointed out and the other one is `--users-per-sec` (Open Model). To read more about closed and open model and how it works in gatling please refer to this [page](https://gatling.io/docs/gatling/reference/current/core/injection/#open-vs-closed-workload-models) from gatling official docs.

> The tester needs a way to specify the pause between refreshes. An argument? 

I think we did something like this in one of the previous attempts of creating a refresh token scenario for a particular situation modeled by @tkyjovsk ., we could follow a similar pattern to use `--user-think-time` and implement that config inside the specific scenario which enables refreshToken

```
    .openLoginPage(true)

    .userThinkPause()
    .loginUsernamePassword()
    .exchangeCode()

    .userThinkPause()
    .refreshToken()

    .userThinkPause()
    .refreshToken()

    .userThinkPause()
    .refreshToken()

    .userThinkPause()
    .logout(true))
```

This is one of the ways, but more than happy to hear your ideas around this implementation.

> The test has to end at some point. Do we have a fixed token refresh count, specified by an arg. Or should it read the --mesurement duration given to the test? And it simply refreshes on the above interval until that time has elapsed?

Your assumption is right about using `--measurement`. That would be the best way to end the test and it takes a default value of 30 seconds which is pointed out in the docs [here](https://www.keycloak.org/keycloak-benchmark/benchmark-guide/latest/configuration#general-parameters).

> The original description talks about 100k users. 

Sorry our docs didn't have a proper way to describe this, but I think we can highlight this in the https://www.keycloak.org/keycloak-benchmark/benchmark-guide/latest/preparing-keycloak section somehow. We do have a dataset module related [documentation](https://www.keycloak.org/keycloak-benchmark/dataset-guide/latest/), and in there we specify how to generate large amounts of datasets. For example to generate a realm with 100k users, you use something like the below command once the dataset module is installed on the target Keycloak instance's providers directory.

```
https://keycloak-example-host.com/realms/master/dataset/create-realms?realm-name=realm-0&count=1&threads-count=10&users-per-realm=100000
```

And the benchmark scenarios are built in such a way when you do not specify a specific user or a client, it anticipates a specific pattern of clients, users and realms to be existing inside the system and runs the requests off of those patterns - see [here](https://github.com/keycloak/keycloak-benchmark/blob/main/benchmark/src/main/scala/keycloak/scenario/KeycloakScenarioBuilder.scala#L74) and [here](https://github.com/keycloak/keycloak-benchmark/blob/main/benchmark/src/main/scala/keycloak/scenario/KeycloakScenarioBuilder.scala#L92). This allows you to scale realistically for large datasets and makes the test conditions field like

> Finally, I've already discovered with the minimal test that it is, that it is really not hard at all to crash keycloak.
We are always learning new ways to trouble Keycloak and would be more than happy to hear it via a new GH Issue discussion and see if thats something we have already encountered or not. Again thanks for the continued interest in improving the health of Keycloak product in general. Cheers !

**Comment by ddurham2** [comment]
Concerning the crash.. I think that it may just be minikube shutting it down due to the failed liveness probes under load.  So, false alarm I think.

**Comment by ddurham2** [comment]
I've locally hit 100k users refreshing.   But granted, different systems will handle it differently.

I spun it up in minikube, with 6 pods limited to 2 cores and 4GB RAM, and running it as follows works for me:
- created 100k users in realm-0 using the [dataset provider](https://www.keycloak.org/keycloak-benchmark/dataset-guide/latest/using-provider)
- kcb.sh --scenario=keycloak.scenario.authentication.AuthCodeWithRefreshToken --server-url=$KC_URL --realm-name=realm-0 --ramp-up=500 --users-per-realm 100000 --user-think-time=100 measurement=600 --concurrent-users=100000

```
================================================================================
2023-07-29 01:19:56                                        1718s elapsed
---- Requests ------------------------------------------------------------------
> Global                                                   (OK=1400000 KO=0     )
> Browser to Log In Endpoint                               (OK=100000 KO=0     )
> Browser posts correct credentials                        (OK=100000 KO=0     )
> Exchange Code                                            (OK=100000 KO=0     )
> RefreshToken                                             (OK=1000000 KO=0     )
> Browser logout                                           (OK=100000 KO=0     )

---- Authentication - Authorization Code Username/Password with Refresh Token --
          active: 0      / done: 100000
================================================================================

Simulation keycloak.scenario.authentication.AuthCodeWithRefreshToken completed in 1718 seconds
Parsing log file(s)...
Parsing log file(s) done
Generating reports...

================================================================================
---- Global Information --------------------------------------------------------
> request count                                    1400000 (OK=1400000 KO=0     )
> min response time                                      2 (OK=2      KO=-     )
> max response time                                  34309 (OK=34309  KO=-     )
> mean response time                                  4509 (OK=4509   KO=-     )
> std deviation                                       7873 (OK=7873   KO=-     )
> response time 50th percentile                        328 (OK=328    KO=-     )
> response time 75th percentile                       3392 (OK=3386   KO=-     )
> response time 95th percentile                      23990 (OK=23989  KO=-     )
> response time 99th percentile                      27971 (OK=27971  KO=-     )
> mean requests/sec                                814.427 (OK=814.427 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                        822712 ( 59%)
> 800 ms < t < 1200 ms                               61965 (  4%)
> t > 1200 ms                                       515323 ( 37%)
> failed                                                 0 (  0%)
================================================================================
```
... not always the best response times.

Do we want to change anything about the actual testing of [the refresh operation itself](https://github.com/keycloak/keycloak-benchmark/pull/434/files#diff-6cd77690a1aaa2dc4d2aef1b73a8137e47e08345dab512836362aa6a9966070aR746)?  If not, then, how to move this across the finish line...
- How do we want to specify the [number of refreshes](https://github.com/keycloak/keycloak-benchmark/pull/434/files#diff-6cd77690a1aaa2dc4d2aef1b73a8137e47e08345dab512836362aa6a9966070aR748) to make?
  -  We _are_ using userThinkTime between refreshes.  I don't love it, but it works.
- During the test I'm not sure it was really using all of the 100k users I created.  In the admin console I could only see that `user-0` had session created (how many it doesn't display).  But `user-1` had none.  It has been my understanding that specifying `--users-per-realm` would cause it to authenticate, cycling through these users.  Is that correct?
- Is that final report it spits out adequate?  It doesn't seem to focus just on the refresh times which is what we're more interested in.  Is it always just a general report?
- Do we want to augment the documentation to mention this scenario?  I assume doc/dataset/modules/ROOT/pages/ is the correct place to make these changes?  I'll take a stab at explaining it.
  - I'll put appropriate links and instructions for using the dataset provider, since it's pretty helpful in this scenario.

**Comment by ddurham2** [comment]
.. and for kicks I did some more runs.  A pod given 12 CPUs and only 4GB of RAM appears to perform better than the 6 pod configuration above (I ran the 6 pod configuration again and got the same poorer results).  The single pod configuration's CPU actually hovered around 5-6 cores utilized and about topped out around 1.5GB RAM.  I'm sure there's many factors, but I thought it worth mentioning.  

Results of 1 Pod w/ 12 CPU and 4GB limit 
```
================================================================================
2023-07-29 02:11:13                                        1610s elapsed
---- Requests ------------------------------------------------------------------
> Global                                                   (OK=1400000 KO=0     )
> Browser to Log In Endpoint                               (OK=100000 KO=0     )
> Browser posts correct credentials                        (OK=100000 KO=0     )
> Exchange Code                                            (OK=100000 KO=0     )
> RefreshToken                                             (OK=1000000 KO=0     )
> Browser logout                                           (OK=100000 KO=0     )

---- Authentication - Authorization Code Username/Password with Refresh Token --
          active: 0      / done: 100000
================================================================================

Simulation keycloak.scenario.authentication.AuthCodeWithRefreshToken completed in 1610 seconds
Parsing log file(s)...
Parsing log file(s) done
Generating reports...

================================================================================
---- Global Information --------------------------------------------------------
> request count                                    1400000 (OK=1400000 KO=0     )
> min response time                                      0 (OK=0      KO=-     )
> max response time                                   2570 (OK=2570   KO=-     )
> mean response time                                    19 (OK=19     KO=-     )
> std deviation                                         50 (OK=50     KO=-     )
> response time 50th percentile                          4 (OK=4      KO=-     )
> response time 75th percentile                          9 (OK=9      KO=-     )
> response time 95th percentile                         99 (OK=99     KO=-     )
> response time 99th percentile                        201 (OK=201    KO=-     )
> mean requests/sec                                869.025 (OK=869.025 KO=-     )
---- Response Time Distribution ------------------------------------------------
> t < 800 ms                                       1399335 (100%)
> 800 ms < t < 1200 ms                                 583 (  0%)
> t > 1200 ms                                           82 (  0%)
> failed                                                 0 (  0%)
================================================================================
```

**Comment by kami619** [comment]
> kcb.sh --scenario=keycloak.scenario.authentication.AuthCodeWithRefreshToken --server-url=$KC_URL --realm-name=realm-0 --ramp-up=500 --users-per-realm 100000 --user-think-time=100 measurement=600 --concurrent-users=100000

I am still going through all the changes you did but one thing which popped out at me was the missing = for 'users-per-realm 100000' I believe if its passed correctly like this 'users-per-realm=100000' the test should rotate the logins between 100k users.

**Comment by ddurham2** [comment]
Ah! I didn't realize that mattered.  But looking at the sh script, I see it does.  I will correct it (and in the comment).
----
After correcting the argument, results are still similar.  So that's good.

**Comment by ddurham2** [comment]
@kami619, for the refresh token frequency, I found that there's an existing `--refresh-token-period=<seconds>` [config value](https://github.com/keycloak/keycloak-benchmark/blob/main/benchmark/src/main/java/org/keycloak/benchmark/Config.java#L133) already handled.  It's used by the [AdminConsoleScenarioBuilder.scala](https://github.com/keycloak/keycloak-benchmark/blob/main/benchmark/src/main/scala/keycloak/scenario/_private/AdminConsoleScenarioBuilder.scala#L603).

Do you think that's more appropriate than --user-think-time?  If so, I can change it.

**Comment by ddurham2** [comment]
For now, I've introduced --refresh-token-count=<count> to parameterized the number of refreshes.  Let me know if that's cool or not..

**Comment by ddurham2** [comment]
@kami619, it occurs to me now, that if one were to specify `--refresh-token-count=0` to this new scenario, then it's rather equivalent to the existing AuthorizationCode scenario.
Any reason not to simply:
- introduce --refresh-token-count=0 (default)
- use --refresh-token-period for the time between refreshes (or something new if you disagree with reusing that existing arg).   The point of this is so that user-think-time can continue to specify how long after opening the login page it waits.
- make the AuthorizationCode scenario call refresh() with these parameters
- document the new available option in the AuthorizationCode docs

Sound more reasonable than introducing a new scenaior?

**Comment by kami619** [comment]
> @kami619, for the refresh token frequency, I found that there's an existing `--refresh-token-period=<seconds>` [config value](https://github.com/keycloak/keycloak-benchmark/blob/main/benchmark/src/main/java/org/keycloak/benchmark/Config.java#L133) already handled. It's used by the [AdminConsoleScenarioBuilder.scala](https://github.com/keycloak/keycloak-benchmark/blob/main/benchmark/src/main/scala/keycloak/scenario/_private/AdminConsoleScenarioBuilder.scala#L603).
> 
> Do you think that's more appropriate than --user-think-time? If so, I can change it.

You are correct, using refresh-token-period cli input is more appropriate for this than using the user-think-time, please make that change.

> that if one were to specify --refresh-token-count=0 to this new scenario, then it's rather equivalent to the existing AuthorizationCode scenario.

I think this is worth the change and merge both scenario's into one, to make code more consice and clean. Thanks for suggesting the change. I am good with that approach.

**Comment by ddurham2** [comment]
all changes complete as far as I'm concerned

**Comment by kami619** [comment]
@ddurham2 thanks much, I will give it another go and merge if everything lines up.
